### DISCUSSION TITLE
`zopen build` should automatically download dependencies

### DISCUSSION THREAD

**Discussion by AnthonyGiorgio** [zopencommunity/meta#48] [Ideas]
When trying to build a port, `zopen build` checks the dependency list to make sure that all the prerequisites are satisfied.  What I find odd is that it just errors out when one is missing. Why couldn't it just download the missing dependency for you?  It already has the ability to do so via `zopen download`, and it knows which dependencies are needed.

This is something that Gentoo's ebuild system was able to do 20 years ago, to great effect.  We don't need something that complicated, but a simple bit of scripting would make life easier for new users.   Plus it will look really cool when something like `git` gets automatically built with one command.

**Comment by IgorTodorovskiIBM** [comment]
I think this is a good idea!

A couple of things we should probably consider if we decide to go this route:
* If the dependency already exists in the pre-defined $HOME/zopen/prod location but a newer version of the dependency is found on github, do we overwrite it? Should we only overwrite only if the "build health" is equal or greater? 
* OR should we support something like version ranges (as in npm: https://github.com/npm/node-semver#versions)? So if perl>=5.3.5 is specified, as long as you have that in version of perl in your path, then you are good.

Also, some of the deps are currently from Rocket software (like git) which cannot be downloaded via zopen download. However, git is a requirement in order to clone a repo so we may be ok.

**Reply by AnthonyGiorgio** [reply]
I wouldn't automatically upgrade a dependency. We could certainly have a `zopen download -u` which would refresh everything, but I don't think that's necessary for a build.  
I do think version numbers are an eventual requirement here.  It's going to complicate things, unfortunately, but it's going to be necessary for proper builds.  The logic to get this right is going to be tricky.

**Comment by MikeFultonDev** [comment]
done
